core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +284,116 @@ public void segmentsLost(IntSet lostSegments) {
-                     // This means the segment was lost in the middle of processing
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.completedNull();
-                     }
-                     return CompletableFuture.completedFuture(value);
-                  });
-               }).sequential();
+   /**
+    * This method iteratively submits a task to operate on the cpu bound thread pool for each segment that is available.
+    * The tasks perform a type of work stealing where they attempt to retrieve the next available segment and process
+    * them as fast as possible. It is entirely possible that a given task is never submitted due to the other tasks
+    * completing all the segments asynchronously. After the main thread has submitted all the tasks it will attempt
+    * to steal a segment and run it if possible and if it can will subsequently attempt to complete all remaining
+    * segments in the same fashion as the other threads. NOTE that this behavior is not normally how reactive
+    * streams are done as given operations are not normally performed until the returned Flowable is subscribed to, but
+    * for performance reasons this method eagerly publishes entries.
+    * <p>
+    * The results of each segment data will then be published each as a single result in the returned Flowable. Due
+    * to the results being retrieved eagerly it is entirely possible that if the Subscriber of the Flowable is slow
+    * that that results queue up. But due to that the fact that results are reduced to single values for each segment
+    * this shouldn't become an issue.
+    * @param set CacheSet to retrieve the publisher for (non-nullable)
+    * @param keysToExclude whether given keys should be excluded from the processing (nullable)
+    * @param toKeyFunction function to convert an entry to a key to determine if it is excluded (must be non null if keysToExclude is)
+    * @param segments the segments to process results for (non-nullable)
+    * @param transformer reducer to collate all the entries for a given segment into a single result (non-nullable)
+    * @param listener listener that handles segments being lost and determining what results should be discarded (non-nullable)
+    * @param concurrentSegments segments map of semgnets left to complete. remove an entry when a segment is completed to
+    *                           prevent a data rehash causing a retry for the given segment
+    * @param <I> input type of the data
+    * @param <R> resulting value
+    * @return Flowable that publishes a result for each segment
+    */
+   protected <I, R> Flowable<R> exactlyOnceParallel(CacheSet<I> set,
+         Set<K> keysToExclude, Function<I, K> toKeyFunction, IntSet segments,
+         Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         SegmentListener listener, IntSet concurrentSegments) {
+      int extraThread = cpuCount;
```

Comment:
I don't agree. The user wanted a parallel publisher, so that is what we give them. They can use a sequential one if they need be. Also these operations are all non blocking, so they should be done relatively quickly and OS scheduler can deschedule still if needed.

Reply:
Is there a rule for how many threads a parallel publisher is supposed to use, or could we use just 2 threads for all the parallel publishers and still call them parallel publishers? :)

I haven't done any performance testing, but IMO putting all the tasks together in the same CPU thread pool and giving them the same priority only works if all the tasks use about the same amount of CPU. Since key-based operation process one single entry, and publisher tasks process millions of entries (there isn't even one task per segment), it's really easy for a few parallel publishers to hog all the threads and delay individual key operations a long time.

Reply:
> Is there a rule for how many threads a parallel publisher is supposed to use, or could we use just 2 threads for all the parallel publishers and still call them parallel publishers? :)

I would think most people would believe that it would similarly to standard parallel Java Streams which is to use all available CPUs. (remember distributed streams uses publisher underneath).

> I haven't done any performance testing, but IMO putting all the tasks together in the same CPU thread pool and giving them the same priority only works if all the tasks use about the same amount of CPU. Since key-based operation process one single entry, and publisher tasks process millions of entries (there isn't even one task per segment), it's really easy for a few parallel publishers to hog all the threads and delay individual key operations a long time.

I don't know what you mean by the fact that there isn't even on task per segment... as there is really up to numCPU tasks that work on segments. In my testing with 200K elements (on each node - only looks at 100K due to numOwners=2) most stream operations are completed before even submitting all the tasks. That is one of the benefits of work stealing like this.

Reply:
> I would think most people would believe that it would similarly to standard parallel Java Streams which is to use all available CPUs. (remember distributed streams uses publisher underneath).

@wburns I would say the main difference between our streams/publishers and the standard Java Streams is that we can expect much bigger streams, e.g. with off-heap. (I'm not going to mention stores because those probably won't run on the CPU thread pool.)

> I don't know what you mean by the fact that there isn't even on task per segment... as there is really up to numCPU tasks that work on segments. In my testing with 200K elements (on each node - only looks at 100K due to numOwners=2) most stream operations are completed before even submitting all the tasks. That is one of the benefits of work stealing like this.

Ok, but what about when you get to 10 million entries, and each of those entries has to be unmarshalled from off-heap, and then you we also expose the publishers through the API so we run user code for each entry?

I was convinced the standard Java Streams call `Spliterator.split` until they get to a certain batch size, but looking in their source code I found this instead:

http://hg.openjdk.java.net/jdk/jdk11/file/e0041b182e31/src/java.base/share/classes/java/util/stream/AbstractTask.java#l155

IMO we should over-partitioning as well, so if there's nothing else running we use all the CPUs, but if the cache is also doing other operations they also get a chance to run in between the publisher tasks, not just after the entire data container has been published. A `ForkJoinPool` and its version of work stealing may help to keep an iteration on the same thread/cpu under load instead of always bouncing from one thread to another, but I believe the task itself is too low-level to do the work stealing.

Reply:
> @wburns I would say the main difference between our streams/publishers and the standard Java Streams is that we can expect much bigger streams, e.g. with off-heap. (I'm not going to mention stores because those probably won't run on the CPU thread pool.)

There is no difference between amount of data seen between Publisher and Streams.. They would both touch the same data, maybe I am not understanding what you mean here.

> Ok, but what about when you get to 10 million entries, and each of those entries has to be unmarshalled from off-heap, and then you we also expose the publishers through the API so we run user code for each entry?
> 
> I was convinced the standard Java Streams call `Spliterator.split` until they get to a certain batch size, but looking in their source code I found this instead:
> 
> http://hg.openjdk.java.net/jdk/jdk11/file/e0041b182e31/src/java.base/share/classes/java/util/stream/AbstractTask.java#l155
> 
> IMO we should over-partitioning as well, so if there's nothing else running we use all the CPUs, but if the cache is also doing other operations they also get a chance to run in between the publisher tasks, not just after the entire data container has been published. A `ForkJoinPool` and its version of work stealing may help to keep an iteration on the same thread/cpu under load instead of always bouncing from one thread to another, but I believe the task itself is too low-level to do the work stealing.

We already overpartition in comparison to # of cpus unless number of segments are too low. Lets check some numbers. Take a more unlikely example: a 10 node cluster with 256 segments, each node would own approximately 25.6 segments. Now if each node is rather large having 12 core w/ HT. This means each CPU would own ~ 1 segment `256 / (10 * 12)`. I would be very surprised to see a customer with 10 12 node boxes. My guess is it would be more like 5 nodes with 6 core (w/ HT) which is `256 / ( 5 * 12 )` which is still 4.266^ segments per "CPU". From my limited experience even this is a bit overkill CPU wise.

Also to take the inverse case, if we under partition in the case of having too few segments, this means with parallel publisher it wouldn't use all the CPUs and do your other argument. To be honest I am not sure what your stance is anymore :)

Also another MAJOR reason to keep partitioning only at the segment level is that we can very easily handle when a state transfer occurs and marking a segment correctly and throwing out said results. If we started partitioning the Map further we would have to synchronize results after all partitions of a given segment are done before returning even a single result and verify that the segment was not suspected. This would complicate the code substantially.

I think it comes down to if a user wants finer grained tasks they can always increase the # of segments as well.

Reply:
Sorry, I missed this comment @wburns 

> We already overpartition in comparison to # of cpus unless number of segments are too low

> Also to take the inverse case, if we under partition in the case of having too few segments, this means with parallel publisher it wouldn't use all the CPUs and do your other argument. To be honest I am not sure what your stance is anymore :)

My stance is that we don't submit one task per segment at the moment: we only submit `numCpus - 1` tasks that process all the segments. Sure, the tasks don't have a fixed set of segments, but the number of segments is irrelevant: they still hog `numCpus - 1` worker threads until all but `numCpus - 2` segments have been processed.

Assuming these tasks will eventually run on a "cpu thread pool" that only has `numCpus` threads to reduce context switches, and the submitter is also a "cpu thread", this means one publisher could block any other operations that needs a cpu thread.

IMO we don't need the work stealing part: when a task finishes processing a segment, it should not immediately start processing the next available segment, instead it should submit a new task to the executor to allow other operations to run. Submitting `numCpus - 1` tasks at the beginning may still be too much, but we'll need experiments and feedback from users to tune that.

> Also another MAJOR reason to keep partitioning only at the segment level is that we can very easily handle when a state transfer occurs and marking a segment correctly and throwing out said results. If we started partitioning the Map further we would have to synchronize results after all partitions of a given segment are done before returning even a single result and verify that the segment was not suspected. This would complicate the code substantially.

I can understand that, a task per segment is good enough until we hear otherwise from users.

> I think it comes down to if a user wants finer grained tasks they can always increase the # of segments as well.

I'd rather not document increasing the number of segments as a way to improve the latency of single-key operations while bulk operations are running, but sure, they will have an option.